### PR TITLE
feat: provide Codeclimate compatible report fromat

### PR DESCRIPTION
**Pull request by tploss**
Provide code quality report as JSON according to Code Climate specificiation. This is need when using the editorconfig-checker in GitLab CI.

More details: #365

### CODE REVIEW

**Review by klaernie — CHANGES_REQUESTED** [review]
Looks generally good and like it adheres to the specification you left in the comment. Thanks for turning that ugly `if...elseif..else` into a `switch` - I totally forgot to think about if that feature is available in golang when writing it.. ;)

However I'd like to see tests for demonstrating it in action and proving correctness.

Furthermore I think you should disable the verbose and debug logging when `config.Format == codeclimate` - else anyone trying to debug will surround the JSON output with unparsable noise.

Also I thought about, if one should maybe turn the codeclimate generation into a method of `ValidationErrors`, which only runs `ConsolidateErrors` and returns the JSON.

In that regard, I was wondering lately, if it would be better to split `FormatErrors` into per-output-format functions, and change their interface towards  `main()` to utilize an io.Writer. But that could easily be another PR once this one is merged.

File: pkg/error/format_codeclimate.go
Comment (tploss) [reply pkg/error/format_codeclimate.go:46]: Totally overlooked this when adapting for the new consolidated lines. Thanks 😃

**Comment by tploss** [comment]
> Looks generally good and like it adheres to the specification you left in the comment. Thanks for turning that ugly `if...elseif..else` into a `switch` - I totally forgot to think about if that feature is available in golang when writing it.. ;)
> 
> However I'd like to see tests for demonstrating it in action and proving correctness.

👍 
Will definitely create some tests. Just have to first understand this *snapshot* system for test data.

> Furthermore I think you should disable the verbose and debug logging when `config.Format == codeclimate` - else anyone trying to debug will surround the JSON output with unparsable noise.

An easier solution from my perspective would be to just log non-standard messages (so debug, error, etc) on stderr instead of stdout. Then one can easily just use `editorconfig-checker --format codeclimate > report.json` but still see all error and debug messages.
Thoughts?

> Also I thought about, if one should maybe turn the codeclimate generation into a method of `ValidationErrors`, which only runs `ConsolidateErrors` and returns the JSON.
> 
> In that regard, I was wondering lately, if it would be better to split `FormatErrors` into per-output-format functions, and change their interface towards `main()` to utilize an io.Writer. But that could easily be another PR once this one is merged.

I was also not really happy with how a single function would grow per added format but did not want to refactor that completely in my PR that just wants to add a single format.
Can we do that as a separate PR afterwards?

**Comment by klaernie** [comment]
> Will definitely create some tests. Just have to first understand this _snapshot_ system for test data.

It's actually fairly easy to use: you just demonstrate calling your code in error_test.go and hand the result of calling `FormatErrors` to `s.MatchSnapshot`. Then run `UPDATE_SNAPS=true go test ./...` so it generates the new snapshot.

**Comment by tploss** [comment]
> It's actually fairly easy to use: you just demonstrate calling your code in error_test.go and hand the result of calling `FormatErrors` to `s.MatchSnapshot`. Then run `UPDATE_SNAPS=true go test ./...` so it generates the new snapshot.

Done 🙂 
I also verified the test by changing the severity constant to "major" and this subsequently failed the test case.

**Comment by tploss** [comment]
The ci is waiting for an approval before running??
I'm not really used to GitHub CI so I don't know if this is normal. @klaernie @mstruebing

**Comment by klaernie** [comment]
Indeed, the CI is waiting to be approved by a maintainer of the repo (so sadly I cannot trigger it).

**Review by klaernie — APPROVED** [review]

**Comment by klaernie** [comment]
Oh, just noticed you overlooked the same places in the README and cmd/main.go argument help as me, which show the available output formats.

**Comment by mstruebing** [comment]
CI already ran so someone needs to have pressed the button.
It's a mechanism to make sure someone who does a PR doesn't break anything or leak any secrets etc.
You could do all kind of bad stuff with Github Actions, this makes sure that at least someone needs to approve the run.

**Comment by tploss** [comment]
> Oh, just noticed you overlooked the same places in the README and cmd/main.go argument help as me, which show the available output formats.

Thanks, I have fixed it and also provided a nicely formatted example output 🙂

File: cmd/editorconfig-checker/main.go
Comment (ccoVeille) [inline cmd/editorconfig-checker/main.go:130]: I would add a comment to explain why you disable the log

Comment (ccoVeille) [inline cmd/editorconfig-checker/main.go:130]: I would like to suggest adding a constant for code climate

This way code that relies on this value could be easily identified

**Review by ccoVeille — APPROVED** [review]

File: cmd/editorconfig-checker/main.go
Comment (klaernie) [reply cmd/editorconfig-checker/main.go:130]: I would take it on me to split `FormatErrors` into multiple independent functions once this PR is merged. Now with a fourth format supported this evergrowing function is too hard to read.

I also wondered if there was a neat way to represent the collection of possible output formats, so that the argument parsing code can discover the possible formats without us having to every time adjust it by hand.

Comment (ccoVeille) [reply cmd/editorconfig-checker/main.go:130]: I would love that

It would help to have a type and a method to list all the formats.

The fact this code has to be maintained is q bit a pain

```go
        flag.StringVar(&c.Format, "format", "default", "specify the output format: default, gcc, github-actions, codeclimate")
```

Comment (klaernie) [reply cmd/editorconfig-checker/main.go:130]: Opened #374 to remind me of the refactoring. 

Reading up on golang, I somehow did not find any mention how to implement essentially an enum type. @ccoVeille do you have seen anything alike in the past?

**Comment by theoludwig** [comment]
I've merged this PR, as it seems like it was ready to go. :smile:

Feel free to open follow up PRs for refactoring.

**Pull request merged**
